[ MB-16172 ] Payment service item review should show same information as SIT Dashboard

mmayo-truss opened the pull request:
## [Jira ticket](https://dp3.atlassian.net/browse/MB-16172)

## Summary

This PR changes the calculations in `payment_request/payment_request_shipments_sit_balance.go` to utilize the functions in `shipment_sit_status.go` that calculate  `totalSITDaysAuthorized`, `totalSITDaysRemaining`, and `totalSITEndDate`. 

Here is a link to how to replicate the issue: https://dp3.atlassian.net/wiki/spaces/~62dff422bc2c449f3d9403fd/pages/2213412919/MB-16172+notes#How-to-replicate-the-issue

## Verification Steps for the Author

These are to be checked by the author.

- [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
- [ ] Have the Jira acceptance criteria been met for this change?

## Verification Steps for Reviewers

These are to be checked by a reviewer.

### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/run-storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/about/application-setup/milmove-local-client/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/about/development/testing)

### How to test

1. Take an approved move and create a SIT service item. The bug in staging is seen on a Destination SIT service item, but origin should cause the same bug. I used move code `SITEXT`. It already has a SIT extension requested and origin SIT service items, so for this one you would want to create Destination SIT service items.
2. Go in as a TOO and approve those service items. 
3. Using Postman or in the prime simulator, create multiple payment requests for the DDASIT service item. Make sure you have non overlapping dates. Here is what that looks like in Postman:
![image](https://github.com/transcom/mymove/assets/110134470/58aed9f0-2f1a-41f8-a41e-c6122127ff66)
4. Log in as the TOO and view the SIT Dashboard (on the Move task order tab). This is what you will use for comparison in the next step. Note the total days remaining and the sit authorized end date: 
![image](https://github.com/transcom/mymove/assets/110134470/7d712515-3f81-4b18-8241-8530ad650621)
5. Log in as a TIO and go to the move in question. There should be multiple review service items buttons. 
![image](https://github.com/transcom/mymove/assets/110134470/46baf08a-6422-41de-8786-e84998cf2615)
6. For each review button you click, you will view a different payment request. The Authorized remaining days and the end date should now match the information from the SIT Dashboard. 
![image](https://github.com/transcom/mymove/assets/110134470/85911ace-2cc3-4672-89d1-18f2ea3c7fc2)



### Frontend

- [ ] There are no aXe warnings for UI.
- [ ] This works in [Supported Browsers and their phone views](https://transcom.github.io/mymove-docs/docs/about/supported-browsers) (Chrome, Firefox, Edge).
- [ ] There are no new console errors in the browser devtools.
- [ ] There are no new console errors in the test output.
- [ ] If this PR adds a new component to Storybook, it ensures the component is fully responsive, OR if it is intentionally not, a wrapping div using the `officeApp` class or custom `min-width` styling is used to hide any states the would not be visible to the user.

### Backend

- [ ] Code follows the guidelines for [Logging](https://transcom.github.io/mymove-docs/docs/about/development/logging).
- [ ] The requirements listed in [Querying the Database Safely](https://transcom.github.io/mymove-docs/docs/backend/guides/golang-guide#querying-the-database-safely) have been satisfied.

### Database

#### Any new migrations/schema changes:

- [ ] Follows our guidelines for [Zero-Downtime Deploys](https://transcom.github.io/mymove-docs/docs/backend/setup/database-migrations#zero-downtime-migrations).
- [ ] Have been communicated to #g-database.
- [ ] Secure migrations have been tested following the instructions in our [docs](https://transcom.github.io/mymove-docs/docs/backend/setup/database-migrations#secure-migrations).

## Screenshots

robot-mymove commented:
<!--
  0 failure: 
  1 warning:  Please add the JI...
  
  
  DangerID: danger-id-Danger;
-->


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>Please add the JIRA issue key to the PR title (e.g. MB-123)</td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 36e677ee6ce80e6dc569614bf46d70041d0cebe2
</p>

ronaktruss reviewed (commented):
Nice work! Functionally, this looks good to me. I haven't had a chance to review the tests yet, but I should be able to do that before my EOD.

When the authorized dates on the payment request is negative we show negative days, but on the sit dashboard we show the word "expired". We may want to check with design to see what is expected here.

Eg.
<img width="352" alt="Screenshot 2023-08-01 at 11 14 08 AM" src="https://github.com/transcom/mymove/assets/84742904/b28f2d65-ade3-4853-be7d-a2d86cba9e74">

<img width="667" alt="Screenshot 2023-08-01 at 11 15 22 AM" src="https://github.com/transcom/mymove/assets/84742904/70235b67-f81b-4dd2-9dc6-89794696d561">

ronaktruss reviewed (approved):
The tests look good to me at a high level. There are some details, that aren't clear to me. Could you provide some context for the following changes:
- the removal of the last test in `pkg/services/payment_request/payment_request_shipments_sit_balance_test.go`. Is there another test that covers something similar or is there a different reason that this is no longer needed. 
- It appears as though the structure of the setup for some of these tests have changed to more accurately reflect how moves are actually created in actual usage, is there something that made these change required?

michaelvaldes-truss commented on pkg/services/mto_shipment/shipment_sit_status.go:
```diff
@@ -151,7 +156,7 @@ func daysInSIT(serviceItem models.MTOServiceItem, today time.Time) int {
 }
 
-func calculateTotalDaysInSIT(shipmentSITs sortedShipmentSITs, today time.Time) int {
+func CalculateTotalDaysInSIT(shipmentSITs SortedShipmentSITs, today time.Time) int {
```
**suggestion(non-blocking):** We might want to consider not passing `today` into a lot of these functions anymore; since they're public now and can be used in multiple places it might make more sense to derive `today` within these calculation functions instead of relying on the caller to supply it.

mmayo-truss replied:
I am not sure. I feel like it is six in one hand/ half a dozen in the other. We would need to calculate `today` in four different functions that way, instead of passing it in.

michaelvaldes-truss commented on pkg/services/mto_shipment/shipment_sit_status.go:
```diff
@@ -89,7 +94,7 @@ func (f shipmentSITStatus) CalculateShipmentSITStatus(appCtx appcontext.AppConte
 		sitEntryDate := *currentSIT.SITEntryDate
 		sitDepartureDate := currentSIT.SITDepartureDate
-		sitAllowanceEndDate := calculateSITAllowanceEndDate(shipmentSITStatus.TotalDaysRemaining, sitEntryDate, today)
+		sitAllowanceEndDate := CalculateSITAllowanceEndDate(shipmentSITStatus.TotalDaysRemaining, sitEntryDate, today)
```
**suggestion(non-blocking):** We might want to consider extracting out all the lines that get info for the current sit into a new function that accepts a sit service item, returns an object containing this sit information (daysInSIT, sitAllowanceEndDate, location, etc.), and can be called from [payment_request_shipments_sit_balance.go](https://github.com/transcom/mymove/pull/11115/files#diff-6b5b0eba82a4abc670c26051304ed31f3dbf1eb328a2ee7045137085c40f1be1), since these calculations are interdependent and are now being called twice.

mmayo-truss replied:
I like this idea. However, with the introduction of partial SIT on the horizon. I am reluctant to spend too much time refactoring, when we will be inevitably changing things in the near future.

michaelvaldes-truss replied:
Gotcha, that makes a lot of sense 👍

michaelvaldes-truss reviewed (commented):
Just leaving a review here with some non-blocking comments since I won't be able to really test this for another day or so.
High-level this looks great! Reusing the SIT calculations seems to have gone really smoothly 😎

mmayo-truss commented:
> The tests look good to me at a high level. There are some details, that aren't clear to me. Could you provide some context for the following changes:
> 
> * the removal of the last test in `pkg/services/payment_request/payment_request_shipments_sit_balance_test.go`. Is there another test that covers something similar or is there a different reason that this is no longer needed.
> * It appears as though the structure of the setup for some of these tests have changed to more accurately reflect how moves are actually created in actual usage, is there something that made these change required?

@ronaktruss 

Answer to Question 1) 
I meant to include a reason as to why I removed that test in the description of the PR. My apologies. That test was testing the days authorized when the field `SITDaysAllowance` is not set on the shipment object is not set. I switched to utilizing the `CalculateShipmentSITAllowance` in `shipment_sit_status.go`. This does not utilize the `SITDaysAllowance` field and instead calculates the allowance based off of the entitlement plus any SIT extensions requested. Since we are not utilizing that field any more, I removed the test.

Answer to Question 2)
The test scenarios do need to be more reflective of how SIT services items would be created in actual use cases. This is due to the following: The payment request SIT balances are calculated when reviewing payment requests made for DDASIT and DOASIT (the additional days SIT) service items. Before these calculations were done using information from the shipment and the D*ASIT service item. However, the calculations done in `shipment_sit_status.go` based off of approved DOPSIT and DDDSIT (delivery SIT) service items, since these service items will have the most accurate entry and departure dates. [You can see that here.](https://github.com/transcom/mymove/blob/36e677ee6ce80e6dc569614bf46d70041d0cebe2/pkg/services/mto_shipment/shipment_sit_status.go#L45-L46). Thus it is important that both the additional days and the delivery service items are present when creating test scenarios.

mmayo-truss commented:
@ronaktruss I made changes to show 'Expired' instead of negative days. 
![image](https://github.com/transcom/mymove/assets/110134470/2b60d9ef-1fff-4dfb-9a88-dff4d0fb6b36)

dinabeigelman reviewed (approved):
Approved once @mmayo-truss updates the copy to "Expired, ended [DD MMM YYYY]"
